shared/src/main/java/com/couchbase/lite/Replicator.java

```diff
@@ -0,0 +1,617 @@
+            dbName = StringUtils.lastPathComponent(remoteURI.getPath());
+        }
+        // replicate against other database
+        else {
+            otherDB = config.getTarget().getDatabase().internal();
+        }
+
+        // If the URL has a hardcoded username/password, add them as an "auth" option:
+        Map<String, Object> options;
+        if (config.getOptions() != null)
+            options = new HashMap<>(config.getOptions());
+        else
+            options = new HashMap<>();
+        String username = URIUtils.getUsername(remoteURI);
+        /* NOTE:
+        // For C4Socket + OkHttp
+        if (username != null && !options.containsKey(kCBLReplicatorAuthOption)) {
+            Map<String, String> auth = new HashMap<>();
+            auth.put(kCBLReplicatorAuthUserName, username);
+            auth.put(kCBLReplicatorAuthPassword, URIUtils.getPassword(remoteURI));
+            if (options == null)
+                options = new HashMap<String, Object>();
+            options.put(kCBLReplicatorAuthOption, auth);
+        }
+        */
+
+        // For CivetWeb
+        if (username != null && username.length() > 0) {
+            String password = URIUtils.getPassword(remoteURI);
+            options.put("headers", getAuthenticatedHeaders(username, password));
+        } else if (options.containsKey(kCBLReplicatorAuthOption)) {
+            Map<String, Object> auth = (Map<String, Object>) options.get(kCBLReplicatorAuthOption);
+            if (auth != null) {
+                username = (String) auth.get(kCBLReplicatorAuthUserName);
+                String password = (String) auth.get(kCBLReplicatorAuthPassword);
+                options.put("headers", getAuthenticatedHeaders(username, password));
+                options.remove(kCBLReplicatorAuthOption);
+            }
+        }
+
+        // Encode the options:
+        byte[] optionsFleece = null;
+        if (options.size() > 0) {
+            FLEncoder enc = new FLEncoder();
+            enc.write(options);
+            try {
+                optionsFleece = enc.finish();
+            } catch (LiteCoreException e) {
+                Log.e(TAG, "Failed to encode", e);
+            }
+            enc.free();
+        }
+
+        // Push / Pull / Continuous:
+        boolean push = isPush(config.getType());
+        boolean pull = isPull(config.getType());
+        boolean continuous = config.isContinuous();
+
+        c4ReplListener = new C4ReplicatorListener() {
+            @Override
+            public void callback(final C4Replicator repl, final C4ReplicatorStatus status, final Object context) {
+                Log.e(TAG, "C4ReplicatorListener.callback() status -> " + status);
+                final Replicator replicator = (Replicator) context;
+                handler.post(new Runnable() {
+                    @Override
+                    public void run() {
+                        if (repl == replicator.c4repl)
```

Comment:
Following is Obj-C implementation. However, as you wrote, checking before the post operation is cheaper. I will update is
```obj-c
static void statusChanged(C4Replicator *repl, C4ReplicatorStatus status, void *context) {
    auto replicator = (__bridge CBLReplicator*)context;
    dispatch_async(replicator->_dispatchQueue, ^{
        if (repl == replicator->_repl)
            [replicator c4StatusChanged: status];
    });
}
```